### ISSUE TITLE
Objects wrongly rejected by quality flags/checks

### ISSUE THREAD

**Issue by genghisken** [genghisken/ps1#25]
Originally reported by: **Stephen Smartt (Bitbucket: [sjsmartt](https://bitbucket.org/sjsmartt), GitHub: Unknown)**

----------------------------------------

http://star.pst.qub.ac.uk/sne/ps13pi/psdb/candidate/1150143171335125800/

PS15anv

Look at the four images on 57167 - two are flagged as good, two are rejected. But they are all real. This hints at a tweak needed in how we quantify the good objects. Need to find out why they failed and if a simple adjustment would let them through. We could be missing objects because of this.  

----------------------------------------
- Bitbucket: https://bitbucket.org/genghisken/ps1/issue/25

**Comment by genghisken** [comment]
*Original comment by* **Ken Smith (Bitbucket: [genghisken](https://bitbucket.org/genghisken), GitHub: [genghisken](https://github.com/genghisken))**:

----------------------------------------

OK.  If non-detections are listed for an object it's because one of the "bad" flags was set in the detection.  The detection is therefore not in the database, but it will be in the CMF file (where I can do a brute force cone search), so I should be able to find out why it wasn't ingested.

**Comment by genghisken** [comment]
*Original comment by* **Ken Smith (Bitbucket: [genghisken](https://bitbucket.org/genghisken), GitHub: [genghisken](https://github.com/genghisken))**:

----------------------------------------

I dug out the two "non-detections" on 57167.  Our current "bad flag" bitmask is this:

```
0001 0000 0000 0011 1011 1100 1000 1000
```

See http://svn.pan-starrs.ifa.hawaii.edu/trac/ipp/wiki/QUBTCSIPPFlags for explanations of each flag.  Any object with any of these flags set is NOT ingested into the database.

In both the CMF files (RINGS.V3.skycell.2057.048.WS.dif.1151530.cmf and RINGS.V3.skycell.2057.049.WS.dif.1151530.cmf) there is a detection recorded at the expected location.  However, in both cases the FLAGS bitmask was set thus:

```
0000 0010 0010 0000 0000 1000 0000 0001
```

In other words, the "suspected defect" flag (bit 11) was set.  I don't know the details of how this flag gets set within IPP.

**Comment by genghisken** [comment]
*Original comment by* **Stephen Smartt (Bitbucket: [sjsmartt](https://bitbucket.org/sjsmartt), GitHub: Unknown)**:

----------------------------------------

So how many objects in a night have this flag "bit 11" set ? 

If it is relatively low, we should turn it off and let the machine learning + eyeballing deal with them. This might be a poorly constrained flag. 

Can you check, on a particular night, how many objects were rejected with this flag compared to the total number that were ingested into the database. I would say if this is at the <10% level, or anywhere near it, we should just turn this particular flag rejection off

**Comment by genghisken** [comment]
*Original comment by* **Ken Smith (Bitbucket: [genghisken](https://bitbucket.org/genghisken), GitHub: [genghisken](https://github.com/genghisken))**:

----------------------------------------

Here are the stats for 31st May, which actually reports the ingestion of the two previous nights of data.

| | |
| ---- | ---- |
| Ingested | 15718100 |
| Rejected (Bad flags) | 17207192 |
| Rejected (Galactic Plane) | 6200491 |

Galactic plane objects are thrown away first.  Of the remaining objects, over 52% are ditched because of "bad flags".

Below is the complete breakdown of all the objects with bad flags set.  The vast majority, over 12 million, are rejected because bits 28 and 15 are set ("Could not be determine size" and "Failed to measure moments").  The fact that bit 0 ("Fit with PSF model" - a "good" flag) is also NOT set (see unfiltered flag list below) implies that these really are bad objects.

On the other hand, bit 11 ("Suspected defect") gets set ON ITS OWN 857216 times (5% of the rejected total), and a further 3 million times in conjunction with other bits - especially bits 13 ("Suspected cosmic ray") and 28 ("Could not be determine size").

Conclusion?  We would only increase the number of ingested objects by a few percent if we turn off the bit 11 ("Suspected defect") flag.  It would be interesting to see how many more "non-detections" are actually because of bit 11.  If it turns out to be a regular occurrence (e.g. both non-detections in this object were bit 11 only), I vote that we turn it off. If it's a spread of a combination of bits, it might not be worth switching off the check.

**Filtered List of Bad Flags (only shows the mask bits that were set)**

```
BAD FLAGS 0001 0000 0000 0011 1011 1100 1000 1000


   271320 0000 0000 0000 0000 0000 0000 1000 0000
      749 0000 0000 0000 0000 0000 0100 0000 0000
       19 0000 0000 0000 0000 0000 0100 1000 0000
   857216 0000 0000 0000 0000 0000 1000 0000 0000
       21 0000 0000 0000 0000 0000 1100 0000 0000
  1403060 0000 0000 0000 0000 0010 1000 0000 0000
       40 0000 0000 0000 0000 0010 1100 0000 0000
  1798287 0001 0000 0000 0000 0000 1000 0000 0000
 12707276 0001 0000 0000 0000 1000 0000 0000 0000
   169204 0001 0000 0000 0000 1000 0000 1000 0000

```


**Unfiltered List of Flags**

```
BAD FLAGS 0001 0000 0000 0011 1011 1100 1000 1000

      50  0000 0000 0000 0000 0000 0000 1000 0000
       3  0000 0000 0000 0000 0000 0001 1000 0000
       6  0000 0000 0000 0000 0000 1000 0000 0000
      16  0000 0000 0000 0000 0010 1000 0000 0000
       8  0000 0000 0001 0000 0000 0000 1000 0000
  181612  0000 0010 0010 0000 0000 0000 1000 0001
    5196  0000 0010 0010 0000 0000 0001 1000 0001
     355  0000 0010 0010 0000 0000 0100 0000 0001
       7  0000 0010 0010 0000 0000 0100 1000 0001
       2  0000 0010 0010 0000 0000 0101 0000 0001
       1  0000 0010 0010 0000 0000 0101 1000 0001
  749194  0000 0010 0010 0000 0000 1000 0000 0001
    9182  0000 0010 0010 0000 0000 1001 0000 0001
      17  0000 0010 0010 0000 0000 1100 0000 0001
 1363474  0000 0010 0010 0000 0010 1000 0000 0001
    3675  0000 0010 0010 0000 0010 1001 0000 0001
      38  0000 0010 0010 0000 0010 1100 0000 0001
     319  0000 0010 0010 0000 0100 0100 0000 0001
       1  0000 0010 0010 0000 0100 0101 0000 0001
    5707  0000 0010 0011 0000 0000 0000 1000 0001
      12  0000 0010 0011 0000 0000 0001 1000 0001
       4  0000 0010 0011 0000 0000 0100 1000 0001
       3  0000 0010 0011 0000 0000 1000 0000 0001
     181  0000 0010 0011 0000 0010 1000 0000 0001
       1  0000 0010 0011 0000 0100 0100 0000 0001
 1779282  0001 0000 0000 0000 0000 1000 0000 0000
   17769  0001 0000 0000 0000 0000 1001 0000 0000
12520213  0001 0000 0000 0000 1000 0000 0000 0000
  157737  0001 0000 0000 0000 1000 0000 1000 0000
  186111  0001 0000 0000 0000 1000 0001 0000 0000
    8626  0001 0000 0000 0000 1000 0001 1000 0000
    1230  0001 0000 0001 0000 0000 1000 0000 0000
       6  0001 0000 0001 0000 0000 1001 0000 0000
     952  0001 0000 0001 0000 1000 0000 0000 0000
    2837  0001 0000 0001 0000 1000 0000 1000 0000
       4  0001 0000 0001 0000 1000 0001 1000 0000
       9  0010 0000 0000 0000 0000 0000 1000 0000
       1  0010 0000 0000 0000 0000 0001 1000 0000
       5  0010 0000 0001 0000 0000 0000 1000 0000
   54850  0010 0010 0010 0000 0000 0000 1000 0001
    2785  0010 0010 0010 0000 0000 0001 1000 0001
       6  0010 0010 0010 0000 0000 0100 0000 0001
       3  0010 0010 0010 0000 0000 0100 1000 0001
       1  0010 0010 0010 0000 0000 0101 0000 0001
   74998  0010 0010 0010 0000 0000 1000 0000 0001
    3457  0010 0010 0010 0000 0000 1001 0000 0001
       2  0010 0010 0010 0000 0000 1100 0000 0001
   12742  0010 0010 0010 0000 0010 1000 0000 0001
     454  0010 0010 0010 0000 0010 1001 0000 0001
       1  0010 0010 0010 0000 0010 1100 0000 0001
      30  0010 0010 0010 0000 0100 0100 0000 0001
    3402  0010 0010 0011 0000 0000 0000 1000 0001
      17  0010 0010 0011 0000 0000 0001 1000 0001
       3  0010 0010 0011 0000 0000 1000 0000 0001
      65  0010 0010 0011 0000 0010 1000 0000 0001
       5  0100 0000 0000 0000 0000 0000 1000 0000
       1  0100 0000 0000 0000 0000 1000 0000 0000
       1  0100 0000 0000 0000 0010 1000 0000 0000
    9401  0100 0010 0010 0000 0000 0000 1000 0001
     345  0100 0010 0010 0000 0000 0001 1000 0001
       5  0100 0010 0010 0000 0000 0100 0000 0001
       3  0100 0010 0010 0000 0000 0100 1000 0001
   17422  0100 0010 0010 0000 0000 1000 0000 0001
     339  0100 0010 0010 0000 0000 1001 0000 0001
       2  0100 0010 0010 0000 0000 1100 0000 0001
   21058  0100 0010 0010 0000 0010 1000 0000 0001
     107  0100 0010 0010 0000 0010 1001 0000 0001
       1  0100 0010 0010 0000 0010 1100 0000 0001
      25  0100 0010 0010 0000 0100 0100 0000 0001
     702  0100 0010 0011 0000 0000 0000 1000 0001
      11  0100 0010 0011 0000 0010 1000 0000 0001
       1  0110 0000 0001 0000 0000 0000 1000 0000
    6472  0110 0010 0010 0000 0000 0000 1000 0001
     268  0110 0010 0010 0000 0000 0001 1000 0001
       2  0110 0010 0010 0000 0000 0100 0000 0001
       1  0110 0010 0010 0000 0000 0100 1000 0001
    2472  0110 0010 0010 0000 0000 1000 0000 0001
     139  0110 0010 0010 0000 0000 1001 0000 0001
    1245  0110 0010 0010 0000 0010 1000 0000 0001
      27  0110 0010 0010 0000 0010 1001 0000 0001
       2  0110 0010 0010 0000 0100 0100 0000 0001
     469  0110 0010 0011 0000 0000 0000 1000 0001
       4  0110 0010 0011 0000 0010 1000 0000 0001
```
